packages/daemon/src/service/pipeline.ts

```diff
@@ -273,25 +273,45 @@ export class PipelineService {
         workingDir: runnable.workingDir,
         template: 'node' // set node by default
       });
-    } catch (err) {
-      job.status = PipelineStatus.FAIL;
-      job.error = err.message;
+
+      // update job status to successful
+      const result = await runnable.valueOf(output) as EvaluateResult;
+      const datasetVal = await runnable.valueOf(dataset) as UniDataset;
+      if (datasetVal?.metadata) {
+        job.dataset = JSON.stringify(datasetVal.metadata);
+      }
+      job.evaluateMap = JSON.stringify(result);
+      job.evaluatePass = result.pass;
+      job.endTime = Date.now();
+      job.status = PipelineStatus.SUCCESS;
       await job.save();
+    } catch (err) {
+      if (!runnable.canceled) {
+        job.status = PipelineStatus.FAIL;
+        job.error = err.message;
+        await job.save();
+      }
       throw err;
     } finally {
       await runnable.destroy();
       delete runnableMap[job.id];
     }
   }
 
-  stopJob(id: string): boolean {
-    const runnable = runnableMap[id];
-    if (runnable) {
-      runnable.destroy();
-      delete runnableMap[id];
-      return true;
+  async stopJob(id: string): Promise<void> {
+    const job = await this.getJobById(id);
+    if (job && job.status === PipelineStatus.RUNNING) {
+      const runnable = runnableMap[id];
+      if (runnable) {
+        runnable.destroy();
+        delete runnableMap[id];
+      } else {
+        console.error(`no runnable found: ${id}`);
+      }
+      job.status = PipelineStatus.CANCELED;
+      await job.save();
     } else {
-      return false;
+      this.ctx.throw(HttpStatus.BAD_REQUEST, 'job is not running');
```

Comment:
Service shouldn't perceive the controller stuff, just throw an error to the enclosing controller

Reply:
But how could throw a bad request error?

Reply:
All right, we can't distinguish it with internal errors now without custom error, but it's not a good start since you can't find a typical response clearly in the controller strategy. Maybe this problem can be put together with the improvement of the global error handle at another PR.

Reply:
Maybe we can throw a [HttpError](https://www.npmjs.com/package/http-errors) which is the same as `ctx.throw` throw.